### PR TITLE
JAMES-2578 Introduce Attribute abstraction

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.mailet;
+
+import java.util.Optional;
+
+public interface QueueSerializable {
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java]: What is a `Queue*` stuff doing in the mailet API?

Maybe we should try to generify that name...

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.LongNode;
+import com.fasterxml.jackson.databind.node.NumericNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+/** 
+ * Serializer
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/Serializer.java]: What does it serialize?

Which benefits does it brings from Java deserialisation?

File: mailet/api/src/main/java/org/apache/mailet/AttributeName.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+/** 
+ * Strong typing for attribute name, which represent the name of an attribute store in a mail.
+ * 
+ * @since Mailet API v3.2
+ */
+public class AttributeName {
+    private final String name;
+
+    public static AttributeName of(String name) {
+        return new AttributeName(name);
```
Comment (chibenwa) [inline mailet/api/src/main/java/org/apache/mailet/AttributeName.java]: Are all attribute names valid? `""` or `null`?

Can you add edge cases for this?

File: mailet/api/src/test/java/org/apache/mailet/AttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class AttributeTest {
+
+    @Test
+    void shouldRespectBeanContract() {
+        EqualsVerifier.forClass(Attribute.class).verify();
+    }
+
```
Comment (chibenwa) [inline mailet/api/src/test/java/org/apache/mailet/AttributeTest.java]: Can the `AttributeName` or the `AttributeValue` be null?

File: mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class AttributeValueTest {
+    @Test
+    void shouldRespectBeanContract() {
+        EqualsVerifier.forClass(AttributeValue.class).verify();
+    }
```
Comment (chibenwa) [inline mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java]: Can an underlying value be `null`? (IMO it should be allowed...)

File: mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+    @Test
+    void stringShouldBeSerializedAndBack() {
```
Comment (chibenwa) [inline mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java]: For each serialiser add test for :
 - `null`
 - Try to pass the json value generated by a serializer to another serializer to exercice error handling.

File: mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+        JsonNode json = expected.toJson();
+        AttributeValue<?> actual = AttributeValue.fromJson(json);
+
+        assertThat(actual).isEqualTo(expected);
+    }
+
+    @Test
+    void doubleShouldBeSerializedAndBack() {
+        AttributeValue<Double> expected = AttributeValue.of(1.0);
```
Comment (chibenwa) [inline mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java]: `1.0d`

File: mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+    @Test
+    void fromJsonStringShouldReturnUrlAttributeValueWhenUrl() throws Exception {
+        AttributeValue<URL> expected = AttributeValue.of(new URL("https://james.apache.org/"));
+
+        AttributeValue<?> actual = AttributeValue.fromJsonString("{\"serializer\":\"UrlSerializer\",\"value\": \"https://james.apache.org/\"}");
+
+        assertThat(actual).isEqualTo(expected);
+    }
+
```
Comment (chibenwa) [inline mailet/api/src/test/java/org/apache/mailet/AttributeValueTest.java]: Can you write a test combining a broken JSON and `fromJson` ?

Can you write a from JSON test with an unknown serializer field?

Can you write a fromJson test combining a Text serializer field and an Integer value? (idem for all serializers)

File: mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java (hunk 1 above)
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java]: We definitively have to find a name for that

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        if (value instanceof QueueSerializable) {
+            return of((QueueSerializable) value);
+        }
+        if (value instanceof URL) {
+            return of((URL) value);
+        }
+        if (value instanceof Serializable) {
+            return ofSerializable((Serializable) value);
+        }
+        throw new IllegalArgumentException("input should at least be Serializable");
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: Could you add some information about what kind of objects we support? I think the message could be more helpful

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode tree = objectMapper.readTree(json);
+        return fromJson(tree);
+    }
+
+    public static Optional<AttributeValue<?>> optionalFromJsonString(String json) {
+        try {
+            return Optional.of(fromJsonString(json));
+        } catch (IOException e) {
+            return Optional.empty();
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: do we need to log error here?

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    }
+
+    @VisibleForTesting
+    static AttributeValue<?> fromJson(JsonNode input) {
+        return Optional.of(input)
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: not `ofNullAble()` ?

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+                .filter(ObjectNode.class::isInstance)
+                .map(ObjectNode.class::cast)
+                .flatMap(AttributeValue::deserialize)
+                .map(AttributeValue::ofAny)
+                .orElseThrow(() -> new IllegalStateException("unable to deserialize " + input.toString()));
+    }
+
+    public static Optional<?> deserialize(ObjectNode fields) {
+        return Optional.ofNullable(fields.get("serializer"))
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: I am afraid that `fields` could be null, then `fields.get...` could throws null pointer exception.

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+                && Objects.equals(this.serializer, that.serializer);
+        }
+        return false;
+    }
+
+    @Override
+    public final int hashCode() {
+        return Objects.hash(value, serializer);
+    }
+}
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: declare `toString()` ?

File: mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            return factory;
+        }
+
+        public AttributeValue<?> getValue() {
+            return value;
+        }
+    }
+
+    public interface Factory {
+        Optional<QueueSerializable> deserialize(Serializable serializable);
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java]: A `factory` deserialize a `serializable` to an optional of `serializable`. Seem not fluent. I mean we usually have a `Deserializer` deserialize a `serializable` to something but definitely not a `serializable`

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 15]
```diff
@@ -44,6 +171,11 @@ public T value() {
         return value;
     }
 
+    //FIXME : poor performance
+    public AttributeValue<T> duplicate() {
```
[the remark this answers is not in the document]
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: Not yet

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+
+    class FSTSerializer implements Serializer<Serializable> {
+        @Override
+        public JsonNode serialize(Serializable object) {
+            FSTConfiguration conf = FSTConfiguration.createJsonConfiguration();
+            String json = conf.asJsonString(object);
+            try {
+                return new ObjectMapper().reader().readTree(json);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
```
[the remark this answers is not in the document]
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/Serializer.java]: Thanks eclipse :)

File: mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java (hunk 14 above)
Comment (blackheaven) [reply mailet/api/src/main/java/org/apache/mailet/QueueSerializable.java]: I've take another approach I've implemented at the beginning, it looks like C++, I don't it's more understandable, but at least it's more type correct.

**Comment by chibenwa** [comment]
We should have a serializer for optional, no?

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java [hunk 17]
```diff
@@ -0,0 +1,473 @@
...
+ * Controlled Json serialization/deserialization
+ * 
+ * @since Mailet API v3.2
+ */
+public interface Serializer<T> {
+    JsonNode serialize(T object);
+
+    Optional<T> deserialize(JsonNode json);
+
+    String getName();
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/Serializer.java]: My proposal: you could declare `getName()` like:
```java
    default String getName() {
        return this.getClass().getSimpleName();
    }
```
And then don't need to override it in impls, I see all impls are using their class name for `getName()`

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java [hunk 18]
```diff
@@ -0,0 +1,473 @@
...
+
+    class IntSerializer implements Serializer<Integer> {
+        @Override
+        public JsonNode serialize(Integer object) {
+            return IntNode.valueOf(object);
+        }
+
+        @Override
+        public Optional<Integer> deserialize(JsonNode json) {
+            if (json instanceof NumericNode) {
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/Serializer.java]: `NumericNode` seems not is a tight condition, how about `IntNode`?
And similar to other Serializers

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java [hunk 19]
```diff
@@ -0,0 +1,473 @@
...
+        @SuppressWarnings("unchecked")
+        private Optional<T> deserialize(String serializer, AttributeValue<?> value) {
+            try {
+                Class<?> factoryClass = Class.forName(serializer);
+                if (ArbitrarySerializable.Factory.class.isAssignableFrom(factoryClass)) {
+                    ArbitrarySerializable.Factory<T> factory = (ArbitrarySerializable.Factory<T>) factoryClass.newInstance();
+                    return factory.deserialize(new ArbitrarySerializable.Serializable<>(value, (Class<ArbitrarySerializable.Factory<T>>) factoryClass));
+                }
+            } catch (Exception e) {
+                LOGGER.error("Error while deserializing", e);
```
Comment (trantienduchn) [inline mailet/api/src/main/java/org/apache/mailet/Serializer.java]: could helps us more when logging information about `serializer` and attribute `value` also

File: mailet/api/src/main/java/org/apache/mailet/Attribute.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+package org.apache.mailet;
+
+import java.util.Objects;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+
+/** 
+ * Attribute attached to an email
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailet/api/src/main/java/org/apache/mailet/Attribute.java]: Why? This class describes a single attribute. Why should we have plural here?

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static AttributeValue<Collection<AttributeValue<?>>> of(Collection<AttributeValue<?>> value) {
+        Preconditions.checkNotNull(value, "value should not be null");
+        return new AttributeValue<>(value, new Serializer.CollectionSerializer());
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static AttributeValue<Map<String, AttributeValue<?>>> of(Map<String, AttributeValue<?>> value) {
+        Preconditions.checkNotNull(value, "value should not be null");
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: Sorry, I do not understand this remark.

If it is about extracting the precondition into a new method, extraction of one-liners seems overkill to me.

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java (hunk 17 above)
Comment (chibenwa) [reply mailet/api/src/main/java/org/apache/mailet/Serializer.java]: Hmm I am not super fan of this.

What is wrong with letting people the name they want to give to their serializer?

With your proposal, the name become implicit and I need to know what `getSimpleName` does in order to understand it. I also need to be aware that this default method exist.

I'd rather prefer this name, which is a critical field in the produced JSON, to be explicit.

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java (hunk 12 above)
Comment (chibenwa) [reply mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: I think `fields` being null is a non-sens that should be prevented by our JSON library.

An explicit percondition should be enough IMO.

File: mailet/api/src/main/java/org/apache/mailet/Attribute.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (trantienduchn) [reply mailet/api/src/main/java/org/apache/mailet/Attribute.java]: Oh I thought in a reversing way, that is okay, you are right

File: mailet/api/src/main/java/org/apache/mailet/Serializer.java (hunk 17 above)
Comment (trantienduchn) [reply mailet/api/src/main/java/org/apache/mailet/Serializer.java]: What I propose would helps us in the case, when some one want to change the classes name but forget to update `getName()` method. It's not really the problem. And I am okay with the code, decision is your!

**Comment by rouazana** [comment]
```
[5f1e987234961488f4d0551839f6f55a3e122a57] [ERROR] Errors: 
[5f1e987234961488f4d0551839f6f55a3e122a57] [ERROR]   SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:183 » ConditionTimeout
```
test this please (will end by silent this test...)

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (rouazana) [reply mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: and you prefer to keep it?

**Comment by aduprat** [comment]
```
[INFO] Apache James :: Mailet API ......................... FAILURE [  2.216 s]
...

[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /james-project/mailet/api/src/main/java/org/apache/mailet/Attribute.java:[55,41] cannot find symbol
  symbol:   method duplicate()
  location: variable value of type org.apache.mailet.AttributeValue<?>
[INFO] 1 error
```

File: mailet/api/src/main/java/org/apache/mailet/AttributeName.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+ * Strong typing for attribute name, which represents the name of an attribute stored in a mail.
+ * 
+ * @since Mailet API v3.2
+ */
+public class AttributeName {
+    private final String name;
+
+    public static AttributeName of(String name) {
+        Preconditions.checkNotNull(name, "`name` is compulsory");
+        Preconditions.checkArgument(!name.equals(""), "`name` should not be empty");
```
Comment (aduprat) [inline mailet/api/src/main/java/org/apache/mailet/AttributeName.java]: `Strings.isNullOrEmpty`

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**